### ISSUE TITLE
Export of XGBoost trained on sparse matrix

### ISSUE THREAD

**Issue by petra-rglv** [jpmml/sklearn2pmml#269]
We are working on a pipeline with tf-idf vectorizer and an xgboost classifier. When we try to export this to pmml the predictions with the pipeline **don't match** the predictions with the pmml.

Here is a small example to reproduce.

```
import pandas as pd
import numpy as np
import xgboost
from xgboost import XGBClassifier
print("xgboost: ", xgboost.__version__)
import sklearn2pmml
print("sklearn2pmml: ", sklearn2pmml.__version__)
from sklearn2pmml import sklearn2pmml
from sklearn2pmml import PMMLPipeline
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn2pmml.feature_extraction.text import Splitter

# sample data
corpus = [
    'This is the first document.',
    'This document is the second document.',
    'And this is the third one.',
    'Is this the first document?'] + ['']*200

y = pd.Series(np.array([0,0,1,0] + [0]*200))

# Training
vectorizer = TfidfVectorizer(norm=None, analyzer="word", tokenizer=Splitter(),
                            min_df=1, token_pattern=r"(?u)[^\s]+",
                            max_features=None)

classifier = XGBClassifier(base_score=0.5, booster="gbtree", colsample_bylevel=0.8,
                          colsample_bynode=1, colsample_bytree=1, gamma=0.1, gpu_id=-1,
                          importance_type="gain", interaction_constraints=None, 
                          learning_rate=0.05, max_delta_step=0, max_depth=5,
                          min_child_weight=5, missing=0.0, monotone_constraints=None, 
                          n_estimators=100, n_jobs=-1, nthread=-1, num_parallel_tree=1, 
                          objective="binary:logistic", random_state=42, reg_alpha=0.001,
                          reg_lambda=10, scale_pos_weight=26.4482, seed=42,
                          subsample=0.8, tree_method=None,
                          validate_parameters=False, verbosity=None)


pipeline = PMMLPipeline([
    ("vectorizer", vectorizer),
    ("classifier", classifier)
])

pipeline.fit(pd.Series(corpus, name="input_string"), y)

# Export to pmml
sklearn2pmml(pipeline, "example.pmml", with_repr=True)

# Test pmml
from jpmml_evaluator import make_evaluator
from jpmml_evaluator.py4j import launch_gateway, Py4JBackend
gateway = launch_gateway()
backend = Py4JBackend(gateway)
evaluator = make_evaluator(backend, r"example.pmml", reporting=True).verify()

input_string = 'aaa'
print("input: ", input_string)
print("pipeline result: ", pipeline.predict_proba([input_string]) )
print("evaluator result: ", evaluator.evaluate({"input_string": input_string}))

>>>
xgboost:  1.2.1
sklearn2pmml:  0.66.1
input:  aaa
pipeline result:  [[0.9630605  0.03693949]]
evaluator result:  {'None': 1, 'probability(0)': 0.24201226234436035, 'probability(1)': 0.7579877376556396}
```
​
What we did to reproduce the error is to train the xgboost model on a sparse matrix, this resulted in multiple conditions along the nodes of the trees with this form:
[f4<-9.53674316e-07]
These conditions are then transferred in the pmml as 
`<SimplePredicate field="float(tf-idf(lowercase(input_string), one))" operator="greaterOrEqual" value="-9.536743E-7"/>`

What would be the right way to export this model into pmml?

**Comment by vruusmann** [comment]
I published a new technical article about TF(-IDF) plus XGBoost pipelines just yesterday:
https://openscoring.io/blog/2021/02/27/sklearn_tf_tfidf_xgboost_pipeline/

TL;DR: Your pipeline is incorrect - if you send TF(-IDF) sparse data matrix to XGBoost classifier, then unset cell values are interpreted as missing values, not as zero-count values.

You probably find it hard to believe, but your Python pipeline is making incorrect predictions, whereas (J)PMML is making correct predictions! Read the above technical article, and you shall understand!

**Comment by vruusmann** [comment]
> You probably find it hard to believe, but your Python pipeline is making incorrect predictions

This claim is based on a fact that Python makes a different prediction for sparse vs dense data matrix. Try it yourself, if you insert a `DenseTransformer` step into the pipeline, then Python predictions will change.

(J)PMML always makes the same predictions, no matter if you use sparse or dense data matrix.

This issue is specific to XGBoost. For example, the Scikit-Learn's random forest classifier gives the same prediction for sparse vs dense data matrix also.
